File: android/digia-engage/src/main/java/com/digia/engage/internal/frequency/FrequencyManager.kt
package com.digia.engage.internal.frequency

import com.digia.engage.internal.analytics.KeyValueStore
import org.json.JSONObject

/**
 * Applies frequency capping for natively-rendered campaigns (nudge + survey).
 * Wraps the pure [FrequencyEvaluator] with persistence (one [FrequencyState] per
 * campaignKey in a [KeyValueStore]) and pulls the current analytics sessionId so
 * `session` windows reset exactly when the native session rotates (cold start /
 * 30-min idle / setUserId / clearUserId).
 *
 * @param sessionIdProvider returns the current analytics sessionId, or null when
 *   analytics is disabled (the evaluator then treats session windows as fresh).
 * @param clock injectable for tests; defaults to wall-clock.
 */
internal class FrequencyManager(
    private val store: KeyValueStore,
    private val sessionIdProvider: () -> String?,
    private val clock: () -> Long = { System.currentTimeMillis() },
) {

    /** Gate: whether the campaign may show now. Capping-free policies always pass. */
    fun isAllowed(campaignKey: String, policy: FrequencyPolicy?): Boolean {
        if (policy == null || !policy.hasConstraint()) return true
        val result = FrequencyEvaluator.evaluate(policy, load(campaignKey), clock(), sessionIdProvider())
        return result.allow
    }

    /** Record one show on "Digia Experience Viewed". */
    fun recordShow(campaignKey: String, policy: FrequencyPolicy?) {
        if (policy == null || !policy.hasConstraint()) return
        val next = FrequencyEvaluator.recordShow(policy, load(campaignKey), clock(), sessionIdProvider())
        save(campaignKey, next)
    }

    /** Apply the permanent stop on "Digia Experience Completed" (survey only). */
    fun recordCompleted(campaignKey: String, policy: FrequencyPolicy?) {
        if (policy?.stopOn != STOP_ON_EXPERIENCE_COMPLETED) return
        val prev = load(campaignKey)
        if (prev?.stoppedAt != null) return
        save(campaignKey, FrequencyEvaluator.recordStop(prev, clock()))
    }

    // ── Persistence ───────────────────────────────────────────────────────────

    private fun keyFor(campaignKey: String) = "$KEY_PREFIX$campaignKey"

    private fun load(campaignKey: String): FrequencyState? {
        val raw = store.getString(keyFor(campaignKey), null) ?: return null
        return runCatching {
            val o = JSONObject(raw)
            FrequencyState(
                total = o.optInt("total", 0),
                windowCount = o.optInt("windowCount", 0),
                windowAnchorAt = if (o.has("windowAnchorAt")) o.optLong("windowAnchorAt") else null,
                sessionId = if (o.has("sessionId") && !o.isNull("sessionId")) o.optString("sessionId") else null,
                stoppedAt = if (o.has("stoppedAt")) o.optLong("stoppedAt") else null,
            )
        }.getOrNull()
    }

    private fun save(campaignKey: String, state: FrequencyState) {
        val o = JSONObject()
            .put("total", state.total)
            .put("windowCount", state.windowCount)
        state.windowAnchorAt?.let { o.put("windowAnchorAt", it) }
        state.sessionId?.let { o.put("sessionId", it) }
        state.stoppedAt?.let { o.put("stoppedAt", it) }
        store.putString(keyFor(campaignKey), o.toString())
    }

    companion object {
        const val KEY_PREFIX = "freq:"
        const val STOP_ON_EXPERIENCE_COMPLETED = "experienceCompleted"

        /**
         * Parse the opaque camelCase frequency policy from a campaign's JSON, or
         * null when there is none ("No cap" / inline). Old snake_case payloads
         * parse to a no-constraint policy and are treated as uncapped.
         */
        fun parsePolicy(json: JSONObject?): FrequencyPolicy? {
            if (json == null) return null
            val maxTotal = if (json.has("maxTotal") && !json.isNull("maxTotal")) json.optInt("maxTotal") else null
            val windowJson = json.optJSONObject("maxPerWindow")
            val maxPerWindow = windowJson?.let {
                val count = if (it.has("count") && !it.isNull("count")) it.optInt("count") else return@let null
                val window = it.optString("window", "").takeIf { w -> w.isNotBlank() } ?: return@let null
                FrequencyWindow(count = count, window = window)
            }
            val stopOn = if (json.has("stopOn") && !json.isNull("stopOn")) {
                json.optString("stopOn", "").takeIf { it.isNotBlank() }
            } else {
                null
            }
            val policy = FrequencyPolicy(maxTotal = maxTotal, maxPerWindow = maxPerWindow, stopOn = stopOn)
            return if (policy.hasConstraint()) policy else null
        }
    }
}

File: flutter/lib/api/internal/frequency/frequency_manager.dart
import 'dart:async';
import 'dart:convert';

import '../../../src/preferences_store.dart';
import '../campaign/campaign_model.dart';
import 'frequency_evaluator.dart';
import 'frequency_policy.dart';

/// Applies frequency capping for natively-rendered campaigns (nudge, survey, and
/// guide). Dart port of Android's `FrequencyManager`: wraps the pure
/// [FrequencyEvaluator] with persistence (one [FrequencyState] per campaignKey)
/// and pulls the current session id so `session` windows reset exactly when the
/// session rotates (cold start / idle timeout / user change).
///
/// Persists via [PreferencesStore] (SharedPreferences), whose reads are
/// synchronous once initialized — so the route-time gate ([isAllowed]) stays
/// synchronous. Writes are fire-and-forget; SharedPreferences updates its
/// in-memory cache synchronously, so a later read sees the new value at once.
///
/// [sessionIdProvider] returns the current session id; [clock] is injectable for
/// tests and defaults to wall-clock.
class FrequencyManager {
  static const String _keyPrefix = 'freq.';

  final PreferencesStore _store;
  final String? Function() _sessionIdProvider;
  final int Function() _clock;

  FrequencyManager(
    this._store,
    this._sessionIdProvider, {
    int Function()? clock,
  }) : _clock = clock ?? (() => DateTime.now().millisecondsSinceEpoch);

  /// Gate: whether [campaign] may show now. Capping-free policies always pass.
  bool isAllowed(CampaignModel campaign) => blockReason(campaign) == null;

  /// The reason [campaign] is currently blocked, or `null` when it may show.
  /// `isAllowed` delegates to this so the gate can log *why* it dropped — one
  /// evaluation, no duplicated logic (mirrors Android's `blockReason`).
  FrequencySkipReason? blockReason(CampaignModel campaign) {
    final policy = campaign.frequency;
    if (policy == null || !policy.hasConstraint) return null;
    final result = FrequencyEvaluator.evaluate(
      policy,
      _load(campaign.campaignKey),
      _clock(),
      _sessionIdProvider(),
    );
    return result.allow ? null : result.reason;
  }

  /// Record one show on "Digia Experience Viewed".
  void recordShow(CampaignModel campaign) {
    final policy = campaign.frequency;
    if (policy == null || !policy.hasConstraint) return;
    final next = FrequencyEvaluator.recordShow(
      policy,
      _load(campaign.campaignKey),
      _clock(),
      _sessionIdProvider(),
    );
    _save(campaign.campaignKey, next);
  }

  /// Apply the permanent stop on "Digia Experience Completed" (survey + guide),
  /// only when the policy opted into `stopOn: experienceCompleted`.
  void recordCompleted(CampaignModel campaign) {
    if (campaign.frequency?.stopOn != stopOnExperienceCompleted) return;
    final prev = _load(campaign.campaignKey);
    if (prev?.stoppedAt != null) return;
    _save(campaign.campaignKey, FrequencyEvaluator.recordStop(prev, _clock()));
  }

  // ── Persistence ─────────────────────────────────────────────────────────────

  String _keyFor(String campaignKey) => '$_keyPrefix$campaignKey';

  FrequencyState? _load(String campaignKey) {
    final raw = _store.read<String>(_keyFor(campaignKey));
    if (raw == null) return null;
    try {
      final decoded = jsonDecode(raw);
      if (decoded is Map) {
        return FrequencyState.fromJson(decoded.cast<String, dynamic>());
      }
    } catch (_) {
      // A corrupt record reads as "no state".
    }
    return null;
  }

  void _save(String campaignKey, FrequencyState state) {
    unawaited(_store.write(_keyFor(campaignKey), jsonEncode(state.toJson())));
  }

  /// The only recognised `stopOn` value — permanently stop on completion.
  static const String stopOnExperienceCompleted = 'experienceCompleted';
}
